crates/bifrost/src/providers/replicated_loglet/sequencer/node.rs

```diff
@@ -0,0 +1,236 @@
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{collections::BTreeMap, ops::Deref, sync::Arc};
+
+use super::Error;
+use restate_core::{
+    network::{ConnectionManager, ConnectionSender},
+    Metadata,
+};
+use restate_types::{
+    logs::{LogletOffset, SequenceNumber},
+    replicated_loglet::NodeSet,
+    PlainNodeId,
+};
+use tokio::sync::{watch, Mutex};
+
+#[derive(Debug, Clone)]
+/// sharable in memory log server state.
+pub struct LogServerState {
+    sealed: watch::Sender<bool>,
+    local_tail: watch::Sender<LogletOffset>,
+}
+
+impl Default for LogServerState {
+    fn default() -> Self {
+        Self {
+            sealed: watch::Sender::new(false),
+            local_tail: watch::Sender::new(LogletOffset::OLDEST),
+        }
+    }
+}
+
+impl LogServerState {
+    /// check if log server is sealed or not
+    pub fn is_sealed(&self) -> bool {
+        *self.sealed.borrow()
+    }
+
+    /// seal log server only marks log server as sealed.
+    /// it also notify all waiters on seal
+    pub fn seal(&self) {
+        self.sealed.send_if_modified(|s| {
+            if *s != true {
+                *s = true;
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    /// get current local tail
+    pub fn local_tail(&self) -> LogletOffset {
+        *self.local_tail.borrow()
+    }
+
+    /// update server local tail if and only if new tail is newer
+    /// that last known tail value
+    pub fn maybe_update_local_tail(&self, new_tail: LogletOffset) {
+        self.local_tail.send_if_modified(|m| {
+            if new_tail > *m {
+                *m = new_tail;
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    /// wait for tail to be at this value or higher
+    pub async fn wait_for_tail(&self, value: LogletOffset) -> Option<LogletOffset> {
+        let mut receiver = self.local_tail.subscribe();
+        receiver.wait_for(|v| *v >= value).await.map(|f| *f).ok()
+    }
+
+    /// wait for seal
+    pub async fn wait_for_seal(&self) -> Option<bool> {
+        let mut receiver = self.sealed.subscribe();
+        receiver.wait_for(|v| *v).await.map(|f| *f).ok()
+    }
+}
+
+/// LogServer instance
+#[derive(Clone)]
+pub struct LogServer {
+    node: PlainNodeId,
+    state: LogServerState,
+    sender: ConnectionSender,
+}
+
+impl LogServer {
+    pub fn node(&self) -> &PlainNodeId {
+        &self.node
+    }
+
+    pub fn state(&self) -> &LogServerState {
+        &self.state
+    }
+
+    pub fn sender(&self) -> ConnectionSender {
+        self.sender.clone()
+    }
+}
+
+#[derive(Default)]
+struct LogServerLock(Mutex<Option<LogServer>>);
```

Comment:
Not sure if this needs a newtype. Perhaps you can use type-alias at the top of this file:
```
type LogServerLock = Mutex<Option<RemoteLogServer>>;
```